config.c

```diff
@@ -45,6 +49,150 @@ static uint32_t parse_color(const char *color) {
 	return res;
 }
 
+static int apply_config_option(struct mako_config *config,
+		const char *name, const char *value) {
+	if (strcmp(name, "font") == 0) {
+		free(config->font);
+		config->font = strdup(value);
+		return 0;
+	} else if (strcmp(name, "background-color") == 0) {
+		config->colors.background = parse_color(value);
+		return 0;
+	} else if (strcmp(name, "text-color") == 0) {
+		config->colors.text = parse_color(value);
+		return 0;
+	} else if (strcmp(name, "width") == 0) {
+		config->width = strtol(value, NULL, 10);
+		return 0;
+	} else if (strcmp(name, "height") == 0) {
+		config->height = strtol(value, NULL, 10);
+		return 0;
+	} else if (strcmp(name, "margin") == 0) {
+		config->margin = strtol(value, NULL, 10);
+		return 0;
+	} else if (strcmp(name, "padding") == 0) {
+		config->padding = strtol(value, NULL, 10);
+		return 0;
+	} else if (strcmp(name, "border-size") == 0) {
+		config->border_size = strtol(value, NULL, 10);
+		return 0;
+	} else if (strcmp(name, "border-color") == 0) {
+		config->colors.border = parse_color(value);
+		return 0;
+	} else if (strcmp(name, "markup") == 0) {
+		config->markup = strcmp(value, "1") == 0;
+		return 0;
+	} else if (strcmp(name, "format") == 0) {
+		free(config->format);
+		config->format = strdup(value);
+		return 0;
+	} else if (strcmp(name, "max-visible") == 0) {
+		config->max_visible = strtol(value, NULL, 10);
+		return 0;
+	} else {
+		return 1;
+	}
+}
+
+static bool file_exists(const char *path) {
+	return path && access(path, R_OK) != -1;
+}
+
+static char *get_config_path(void) {
+	static const char *config_paths[] = {
+		"$HOME/.mako/config",
+		"$XDG_CONFIG_HOME/mako/config",
+	};
+
+	if (!getenv("XDG_CONFIG_HOME")) {
+		char *home = getenv("HOME");
+		char *config_home = malloc(strlen(home) + strlen("/.config") + 1);
+		if (!config_home) {
+			fprintf(stderr, "Unable to allocate $HOME/.config\n");
+		} else {
+			strcpy(config_home, home);
+			strcat(config_home, "/.config");
+			setenv("XDG_CONFIG_HOME", config_home, 1);
+			free(config_home);
+		}
+	}
+
+	wordexp_t p;
+	char *path;
+
+	for (int i = 0; i < (int)(sizeof(config_paths) / sizeof(char *)); ++i) {
+		if (wordexp(config_paths[i], &p, 0) == 0) {
+			path = strdup(p.we_wordv[0]);
+			wordfree(&p);
+			if (file_exists(path)) {
+				return path;
+			}
+			free(path);
+		}
+	}
+
+	return NULL; // Not reached
+}
+
+int load_config_file(struct mako_config *config) {
+	char *path = get_config_path();
+	if (!path) {
+		return 0;
+	}
+
+	FILE *f = fopen(path, "r");
+	if (!f) {
+		free(path);
+		return 0;
+	}
+	const char *base = basename(path);
+
+	int lineno = 0;
+	while (!feof(f)) {
+		char *line = NULL;
+		size_t n = 0;
+		ssize_t l = getline(&line, &n, f);
```

Comment:
`getline` can reuse the allocation, so you could get away with a lot few allocations by moving the variables outside of the loop.
```c
size_t l;
char *line = NULL;
size_t n = 0;
while ((l = getline(&line, &n, f)) > 0) {
    // Do stuff
}
free(line);
```